src/qibo/models/iqae.py

```diff
@@ -0,0 +1,384 @@
+import numpy as np
+import scipy.stats
+
+from qibo import gates
+from qibo.config import raise_error
+from qibo.models import Circuit
+
+
+class IQAE:
+    """Model that performs the Iterative Quantum Ampltidue Estimation algorithm.
+
+    The implemented class in this code utilizes the Iterative Quantum Amplitude Estimation (IQAE)
+    algorithm, which was proposed in [1]. The algorithm provides an estimated output that, with a
+    probability `alpha', differs from the target value by `epsilon'. Both `alpha' and `epsilon' can
+    be specified.
+
+    Unlike Brassard's original QAE algorithm [2], this implementation does not rely on Quantum Phase
+    Estimation but instead is based solely on Grover's algorithm. The IQAE algorithm employs a series
+    of carefully selected Grover iterations to determine an estimate for the target amplitude.
+
+    References:
+    [1]: Grinko, D., Gacon, J., Zoufal, C., & Woerner, S. (2019). Iterative Quantum Amplitude Estimation.
+    arXiv:1912.05559 <https://arxiv.org/abs/1912.05559>.
+    [2]: Brassard, G., Hoyer, P., Mosca, M., & Tapp, A. (2000). Quantum Amplitude Amplification and
+    Estimation. arXiv:quant-ph/0005055 <http://arxiv.org/abs/quant-ph/0005055>.
+
+    Args:
+        circuit_A (:class:`qibo.circuit`): quantum circuit that specifies the QAE problem
+        circuit_Q (:class:`qibo.circuit`): quantum circuit of the Grover/Amplification operator
+
+        alpha (float): confidence level, the target probability is 1 - `alpha', has values between 0 and 1
+        epsilon (float): target precision for estimation target `a`, has values between 0 and 0.5
+        method (string): statistical method used to estimate the confidence intervals in
+                each iteration, can be either 'chernoff' (default) for the Chernoff intervals or 'beta' for the
+                Clopper-Pearson intervals
+        n_shots (int): number of shots
+    Raises:
+        ValueError: If epsilon is not in (0, 0.5]
+        ValueError: If alpha is not in (0, 1)
+        ValueError: If method is not supported
+        ValueError: If circuit_A or circuit_Q are not provided or if the number of qubits in circuit_A is
+        greater than in circuit_Q
+
+    Example:
+            import numpy as np
+            from qibo import gates
+            from qibo.models import Circuit
+            from qibo.models.iqae import IQAE
+            ...
+            #Here the circuits A and Q are defined
+            ...
+            iae = IQAE(circuit_A=A, circuit_Q=Q)
+            results=iae.execute()
+            print(results.estimation)
+    """
+
+    def __init__(
+        self,
+        circuit_A=None,
```

Comment:
> Function names should be lowercase, with words separated by underscores as necessary to improve readability.

From https://peps.python.org/pep-0008/#function-and-variable-names ...
```suggestion
        circuit_a=None,
```